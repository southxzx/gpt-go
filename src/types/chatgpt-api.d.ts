interface ChatGPTApiOptions {
  apiKey: string;
  apiURL?: string;
  model?: string;
}

export type { ChatGPTApiOptions };
