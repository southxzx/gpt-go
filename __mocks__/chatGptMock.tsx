import ChatGPT from "../src/api/chatgpt-api";

export const chatGPT = new ChatGPT({
  apiKey: "1234567890",
});

export const mockSendMessage = jest.spyOn(chatGPT, "sendMessage");
mockSendMessage.mockResolvedValueOnce({
  choices: [
    {
      message: {
        content: "welcome to my world",
      },
    },
  ],
});
