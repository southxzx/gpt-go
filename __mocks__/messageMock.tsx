export const mockPostMessage = jest
  .spyOn(window, "postMessage")
  .mockImplementation(() => {
    // window.postMessage()
  });

export const mockOnReceiveMessageGenerate = jest.fn(() => {
  console.log("weeeee");
});

export const mockEventListenerOnMessage = jest
  .spyOn(window, "addEventListener")
  .mockImplementationOnce((event, handler) => {
    if (event === "message") {
      console.log(
        "ðŸš€ ~ file: messageMock.tsx:15 ~ .mockImplementationOnce ~ event:",
        event
      );
      handler = mockOnReceiveMessageGenerate;
    }
  });
