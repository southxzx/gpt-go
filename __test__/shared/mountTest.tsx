import React from "react";
import { render } from "@testing-library/react";

export default function mountTest(Component: React.ComponentType) {
  describe("mount and unmount", () => {
    it("component could be updated and unmounted without errors", () => {
      const { rerender, unmount } = render(<Component />);
      expect(() => {
        rerender(<Component />);
        unmount();
      }).not.toThrow();
    });
  });
}
