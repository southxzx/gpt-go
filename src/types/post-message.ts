export enum POST_MESSAGE_TYPE {
  GENERATE = "generate",
  GENERATE_RESEARCH = "generate-research",
  GET_STORAGE = "get-storage",
  POST_STORAGE = "post-storage",
  GET_SELECTION_TEXT = "get-selection-text",
  RESPONSE_SELECTION_TEXT = "response-selection-text",
  NOTIFY = "notify",
}

export enum MESSAGE_RESPONSE_TEXT {
  SELECT_TEXT = "Please select a text layer to update!",
}
