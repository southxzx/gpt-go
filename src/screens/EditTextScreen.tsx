import get from "lodash/get";
import React, { useEffect } from "react";
import ChatGPTApi from "../api/chatgpt-api";
import Button from "../components/common/Button";
import ItemBox from "../components/common/ItemBox";
import ItemBoxSelect from "../components/common/ItemBoxSelect";
import LoadingLine from "../components/common/LoadingLine";
import TextField from "../components/common/TextField";
import { POST_MESSAGE_TYPE } from "../types/post-message";

const improve_options = ["Simplify", "Make Longer", "Make Shorter"];
const tone_options = [
  "Professional",
  "Confident",
  "Casual",
  "Straightforward",
  "Friendly",
];
const translate_options = [
  "Arabic",
  "Chinese",
  "English",
  "French",
  "German",
  "Japanese",
  "Korean",
  "Portuguese",
  "Russian",
  "Spanish",
  "Thai",
  "Vietnamese",
];

interface IEditTextScreenProps {
  api: ChatGPTApi;
}

const EditTextScreen: React.FC<IEditTextScreenProps> = ({ api }) => {
  const [inputValue, setInputValue] = React.useState<string>("");
  const [quickAccessValue, setQuickAccessValue] = React.useState<string>("");
  const [loading, setLoading] = React.useState<boolean>(false);
  const [needSelectionText, setNeedSelectionText] =
    React.useState<boolean>(false);

  const onGenerate = async () => {
    getSelectionText();
  };

  const getSelectionText = () => {
    parent.postMessage(
      {
        pluginMessage: {
          type: POST_MESSAGE_TYPE.GET_SELECTION_TEXT,
          option: "edit",
        },
      },
      "*"
    );
  };

  const generateResponse = async (selectionValue = "") => {
    setLoading(true);

    const messageValue = selectionValue
      ? `${inputValue || quickAccessValue}: ${selectionValue}`
      : inputValue || quickAccessValue;
    const res = await api.sendMessage(messageValue);
    const message = get(
      res,
      "choices[0].message.content",
      "Can not generate. Please try again or double-check your API Key!"
    );
    parent.postMessage(
      {
        pluginMessage: {
          type: POST_MESSAGE_TYPE.GENERATE,
          message: message,
          needSelectionText,
        },
      },
      "*"
    );
    setLoading(false);
    setQuickAccessValue("");
  };

  const onClickOption = (text: string, key: string) => {
    setQuickAccessValue(text);
    setInputValue("");
  };

  const onReceiveMessageGenerate = (event: MessageEvent<any>) => {
    const { type, message, option } = event.data.pluginMessage;
    if (type === POST_MESSAGE_TYPE.GET_SELECTION_TEXT && option === "edit") {
      message && generateResponse(message);
    }
  };

  const onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (!inputValue) {
      return;
    }
    if (e.key === "Enter") {
      onGenerate();
      e.stopPropagation();
      e.preventDefault();
    }
  };

  useEffect(() => {
    window.addEventListener("message", onReceiveMessageGenerate);
    return () =>
      window.removeEventListener("message", onReceiveMessageGenerate);
  }, [inputValue, quickAccessValue]);

  useEffect(() => {
    if (quickAccessValue) {
      onGenerate();
    }
  }, [quickAccessValue]);

  return (
    <div className="screen-container">
      <div>
        <div className="block-generate-input-wrapper">
          <TextField
            isTextArea={true}
            value={inputValue}
            onChange={(e) => setInputValue(e.target.value)}
            className="block-generate-input"
            placeholder="Generate copy for landing page..."
            onKeyDown={onKeyDown}
            autoFocus={true}
          />
          <Button
            disabled={!inputValue || loading}
            onClick={onGenerate}
            // loading={loading}
            style={{ minWidth: 82 }}
          >
            Generate
          </Button>
        </div>

        <label>Quick Access</label>
        <div className="block-content-wrapper">
          <ItemBoxSelect
            values={tone_options}
            disabled={loading}
            onSelect={(val) => onClickOption(`Change tone to ${val}`, "Tone")}
            placeholder="Change Tone To..."
          />
          <ItemBoxSelect
            values={translate_options}
            disabled={loading}
            onSelect={(val) =>
              onClickOption(`Translate to ${val}`, "Translate")
            }
            placeholder="Translate To..."
          />
          {improve_options.map((option) => (
            <ItemBox
              text={option}
              key={option}
              disabled={loading}
              onClick={() => onClickOption(option, option)}
            />
          ))}
        </div>
      </div>
      {loading && <LoadingLine />}
    </div>
  );
};

export default EditTextScreen;
