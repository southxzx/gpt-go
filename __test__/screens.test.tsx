import React from "react";
import { fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";

import EditTextScreen from "../src/screens/EditTextScreen";

import mountTest from "./shared/mountTest";
import { chatGPT } from "../__mocks__/chatGptMock";
import {
  mockEventListenerOnMessage,
  mockPostMessage,
} from "../__mocks__/messageMock";
import { POST_MESSAGE_TYPE } from "../src/types/post-message";

describe("Screen EditText", () => {
  mountTest(() => <EditTextScreen api={chatGPT} />);

  it("renders correctly", () => {
    const { container } = render(<EditTextScreen api={chatGPT} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should disabled button Generate when input is empty", () => {
    const { container } = render(<EditTextScreen api={chatGPT} />);
    const button = container.querySelector(".btn") as HTMLElement;
    expect(button).toHaveClass("btn-disabled");
  });

  it("should not disabled button Generate when input is not empty", () => {
    const { container } = render(<EditTextScreen api={chatGPT} />);
    const input = container.querySelector(
      ".block-generate-input"
    ) as HTMLElement;
    const button = container.querySelector(".btn") as HTMLElement;
    fireEvent.change(input, { target: { value: "hello" } });
    expect(button).not.toHaveClass("btn-disabled");
  });

  it("should send message to Figma when click Generate", async () => {
    const mockMessage = "hello";
    const { container } = render(<EditTextScreen api={chatGPT} />);
    const input = container.querySelector(
      ".block-generate-input"
    ) as HTMLElement;
    const button = container.querySelector(".btn") as HTMLElement;
    fireEvent.change(input, { target: { value: mockMessage } });
    expect(button).not.toHaveClass("btn-disabled");
    fireEvent.click(button);

    await waitFor(() => {
      expect(mockPostMessage).toHaveBeenCalledTimes(1);
      expect(mockPostMessage).toHaveBeenCalledWith(
        {
          pluginMessage: {
            type: POST_MESSAGE_TYPE.GET_SELECTION_TEXT,
            option: "edit",
          },
        },
        "*"
      );

      expect(mockEventListenerOnMessage).toHaveBeenCalled();
      expect(mockEventListenerOnMessage).toHaveBeenCalledWith(
        "message",
        expect.any(Function)
      );
    });
  });

  it("should remove text when click Quick access option", () => {
    const { container } = render(<EditTextScreen api={chatGPT} />);
    const firstQuickAccessOption = container.querySelector(
      ".item-box-dropdown-wrapper"
    );
    const input = container.querySelector(
      ".block-generate-input"
    ) as HTMLElement;
    fireEvent.change(input, {
      target: {
        value: "hello",
      },
    });

    fireEvent.click(firstQuickAccessOption as HTMLElement);

    const firstQuickAccessOptionItem =
      container.querySelector(".dropdown-option");
    expect(firstQuickAccessOptionItem).toBeInTheDocument();

    fireEvent.click(firstQuickAccessOptionItem?.firstChild as HTMLElement);
    expect(input).toHaveValue("");
  });
});
