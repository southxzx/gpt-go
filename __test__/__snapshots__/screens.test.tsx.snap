// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen EditText renders correctly 1`] = `
<div
  class="screen-container"
>
  <div>
    <div
      class="block-generate-input-wrapper"
    >
      <textarea
        class="block-generate-input"
        placeholder="Generate copy for landing page..."
        rows="3"
      />
      <button
        class="undefined btn btn-disabled"
        disabled=""
        style="min-width: 82px;"
      >
        Generate
      </button>
    </div>
    <label>
      Quick Access
    </label>
    <div
      class="block-content-wrapper"
    >
      <div
        class="dropdown-wrapper"
      >
        <div
          class="item-box-wrapper item-box-dropdown-wrapper  "
        >
          <p>
            Change Tone To...
          </p>
          <svg
            fill="none"
            height="12"
            viewBox="0 0 12 12"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#clip0_110_4423)"
            >
              <rect
                fill="white"
                fill-opacity="0.01"
                height="12"
                width="12"
              />
              <path
                d="M11.9045 3.20086L11.368 2.66433C11.2406 2.53698 11.0347 2.53698 10.9073 2.66433L6 7.5608L1.09269 2.66433C0.965338 2.53698 0.759399 2.53698 0.632042 2.66433L0.0955177 3.20086C-0.0318392 3.32822 -0.0318392 3.53416 0.0955177 3.66151L5.76967 9.33567C5.89703 9.46303 6.10297 9.46303 6.23033 9.33567L11.9045 3.66151C12.0318 3.53416 12.0318 3.32822 11.9045 3.20086Z"
                fill="#20133A"
              />
            </g>
            <defs>
              <clippath
                id="clip0_110_4423"
              >
                <rect
                  fill="white"
                  height="12"
                  width="12"
                />
              </clippath>
            </defs>
          </svg>
        </div>
      </div>
      <div
        class="dropdown-wrapper"
      >
        <div
          class="item-box-wrapper item-box-dropdown-wrapper  "
        >
          <p>
            Translate To...
          </p>
          <svg
            fill="none"
            height="12"
            viewBox="0 0 12 12"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#clip0_110_4423)"
            >
              <rect
                fill="white"
                fill-opacity="0.01"
                height="12"
                width="12"
              />
              <path
                d="M11.9045 3.20086L11.368 2.66433C11.2406 2.53698 11.0347 2.53698 10.9073 2.66433L6 7.5608L1.09269 2.66433C0.965338 2.53698 0.759399 2.53698 0.632042 2.66433L0.0955177 3.20086C-0.0318392 3.32822 -0.0318392 3.53416 0.0955177 3.66151L5.76967 9.33567C5.89703 9.46303 6.10297 9.46303 6.23033 9.33567L11.9045 3.66151C12.0318 3.53416 12.0318 3.32822 11.9045 3.20086Z"
                fill="#20133A"
              />
            </g>
            <defs>
              <clippath
                id="clip0_110_4423"
              >
                <rect
                  fill="white"
                  height="12"
                  width="12"
                />
              </clippath>
            </defs>
          </svg>
        </div>
      </div>
      <div
        class="item-box-wrapper "
      >
        <p>
          Simplify
        </p>
      </div>
      <div
        class="item-box-wrapper "
      >
        <p>
          Make Longer
        </p>
      </div>
      <div
        class="item-box-wrapper "
      >
        <p>
          Make Shorter
        </p>
      </div>
    </div>
  </div>
</div>
`;
